pallets/stateful-storage/src/lib.rs

```diff
@@ -819,31 +1131,174 @@ impl<T: Config> Pallet<T> {
 	pub fn get_paginated_page_for(
 		msa_id: MessageSourceId,
-		schema_id: SchemaId,
+		intent_id: IntentId,
 		page_id: PageId,
 	) -> Result<Option<PaginatedPage<T>>, DispatchError> {
-		let keys: PaginatedKey = (schema_id, page_id);
-		Ok(StatefulChildTree::<T::KeyHasher>::try_read::<_, PaginatedPage<T>>(
+		let keys: PaginatedKey = (intent_id, page_id);
+		let page = StatefulChildTree::<T::KeyHasher>::try_read::<_, PaginatedPage<T>>(
 			&msa_id,
 			PALLET_STORAGE_PREFIX,
 			PAGINATED_STORAGE_PREFIX,
 			&keys,
 		)
-		.map_err(|_| Error::<T>::CorruptedState)?)
+		.map_err(|_| Error::<T>::CorruptedState)?;
+
+		// Note: PageVersion::V1 is unsupported because no pages were ever written with that tag;
+		// hence it indicates some kind of corrupted state. In future, the pallet should be able
+		// to support reading pages with multiple versions.
+		if let Some(Page { page_version, .. }) = &page {
+			if *page_version != PageVersion::V2 {
+				return Err(Error::<T>::UnsupportedPageVersion.into());
+			}
+		};
+		Ok(page)
 	}
 
 	/// Gets an itemized storage for desired parameters
 	pub fn get_itemized_page_for(
 		msa_id: MessageSourceId,
-		schema_id: SchemaId,
+		intent_id: IntentId,
 	) -> Result<Option<ItemizedPage<T>>, DispatchError> {
-		let keys: ItemizedKey = (schema_id,);
-		Ok(StatefulChildTree::<T::KeyHasher>::try_read::<_, ItemizedPage<T>>(
+		let keys: ItemizedKey = (intent_id,);
+		let page = StatefulChildTree::<T::KeyHasher>::try_read::<_, ItemizedPage<T>>(
 			&msa_id,
 			PALLET_STORAGE_PREFIX,
 			ITEMIZED_STORAGE_PREFIX,
 			&keys,
 		)
-		.map_err(|_| Error::<T>::CorruptedState)?)
+		.map_err(|_| Error::<T>::CorruptedState)?;
+
+		// Note: PageVersion::V1 is unsupported because no pages were ever written with that tag;
+		// hence it indicates some kind of corrupted state. In future, the pallet should be able
+		// to support reading pages with multiple versions.
+		if let Some(Page { page_version, .. }) = &page {
+			if *page_version != PageVersion::V2 {
+				return Err(Error::<T>::UnsupportedPageVersion.into());
+			}
+		};
+		Ok(page)
+	}
+}
+
+/// The TransactionExtension trait is implemented on BlockDuringMigration to prevent extrinsics
+/// in this pallet from being executed while a migration is in progress
+#[derive(Encode, Decode, DecodeWithMemTracking, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct BlockDuringMigration<T: Config + Send + Sync>(PhantomData<T>);
+
+impl<T: Config + Send + Sync> BlockDuringMigration<T> {
+	/// Create new `TransactionExtension`
+	pub fn new() -> Self {
+		Self(PhantomData)
+	}
+
+	/// Prevent an extrinsic from being executed while a migraiton is in progress.
+	/// Returns a `ValidTransaction` or wrapped [`ValidityError`]
+	///
+	/// # Errors
+	/// * [`ValidityError::InvalidStorageVersion`] - the on-chain storage version does not match the in-code storage version
+	///
+	pub fn block_extrinsics_during_migration() -> TransactionValidity {
```

Comment:
Q: Does this block all extrinsic calls for Stateful storage during migration? If that is the case how long is our estimate for migration time and we should also consider that we have other semi active providers

Reply:
We have no providers using any Stateful Storage extrinsics since MeWe, AFAIK, since the only stateful storage we currently use on-chain is for Graph, and only MeWe was doing graph stuff.
I'll confirm with @pfrank13 that we're not generating graph keys for current providers.

We'll have a better idea of the migration duration once we run the benchmarks in CI.